src/ring_theory/graded_algebra/basic.lean

```diff
@@ -77,4 +81,98 @@ def graded_algebra.decompose : A ≃ₐ[R] ⨁ i, 𝒜 i := alg_equiv.symm
   (graded_algebra.decompose 𝒜).symm (direct_sum.of _ i x) = x :=
 direct_sum.submodule_coe_alg_hom_of 𝒜 _ _
 
+/-- The projection maps of graded algebra-/
+def graded_algebra.proj (𝒜 : ι → submodule R A) [graded_algebra 𝒜] (i : ι) : A →ₗ[R] A :=
+(𝒜 i).subtype.comp $
+  (dfinsupp.lapply i).comp $
+  (graded_algebra.decompose 𝒜).to_alg_hom.to_linear_map
+
+@[simp] lemma graded_algebra.proj_apply (i : ι) (r : A) :
+  graded_algebra.proj 𝒜 i r = (graded_algebra.decompose 𝒜 r : ⨁ i, 𝒜 i) i := rfl
+
+/-- The support of `r` is the `finset` where `proj R A i r ≠ 0 ↔ i ∈ r.support`-/
+def graded_algebra.support [Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0)]
+  (r : A) : finset ι :=
+(graded_algebra.decompose 𝒜 r).support
+
+lemma graded_algebra.proj_recompose (a : ⨁ i, 𝒜 i) (i : ι) :
+  graded_algebra.proj 𝒜 i ((graded_algebra.decompose 𝒜).symm a) =
+  (graded_algebra.decompose 𝒜).symm (direct_sum.of _ i (a i)) :=
+by rw [graded_algebra.proj_apply, graded_algebra.decompose_symm_of, alg_equiv.apply_symm_apply]
+
+variable [Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0)]
+
+lemma graded_algebra.mem_support_iff
+  (r : A) (i : ι) :
+i ∈ graded_algebra.support 𝒜 r ↔ (graded_algebra.proj 𝒜 i r ≠ 0) :=
+begin
+  rw [graded_algebra.support, dfinsupp.mem_support_iff, graded_algebra.proj_apply],
+  simp only [ne.def, submodule.coe_eq_zero],
+end
+
+lemma graded_algebra.sum_support_decompose (r : A) :
+  ∑ i in graded_algebra.support 𝒜 r, (graded_algebra.decompose 𝒜 r i : A) = r :=
+begin
+  conv_rhs {
+    rw [←(graded_algebra.decompose 𝒜).symm_apply_apply r,
+        ←direct_sum.sum_support_of _ (graded_algebra.decompose 𝒜 r)] },
+  rw [alg_equiv.map_sum, graded_algebra.support],
+  simp_rw graded_algebra.decompose_symm_of,
+end
+
+lemma graded_algebra.mul_proj (r r' : A) (i : ι) :
+  (graded_algebra.decompose 𝒜 (r * r') i : A) =
+  ∑ ij in finset.filter (λ ij : ι × ι, ij.1 + ij.2 = i)
+    ((graded_algebra.support 𝒜 r).product (graded_algebra.support 𝒜 r')),
+    (graded_algebra.decompose 𝒜 r ij.1 : A) * (graded_algebra.decompose 𝒜 r' ij.2 : A) :=
+begin
+  have set_eq : (graded_algebra.support 𝒜 r).product (graded_algebra.support 𝒜 r') =
+  finset.filter (λ ij : ι × ι, ij.1 + ij.2 = i) _ ∪
+  finset.filter (λ ij : ι × ι, ij.1 + ij.2 ≠ i) _ := (finset.filter_union_filter_neg_eq _ _).symm,
+  conv_lhs {
+    rw [←graded_algebra.sum_support_decompose 𝒜 r,
+    ←graded_algebra.sum_support_decompose 𝒜 r', finset.sum_mul_sum, alg_equiv.map_sum, set_eq] },
+  rw finset.sum_union,
+  suffices : ∑ (x : ι × ι) in finset.filter (λ (ij : ι × ι), ij.fst + ij.snd ≠ i)
+    ((graded_algebra.support 𝒜 r).product (graded_algebra.support 𝒜 r')),
+  (graded_algebra.decompose 𝒜
+    ((graded_algebra.decompose 𝒜 r x.fst : A) * (graded_algebra.decompose 𝒜 r' x.snd : A)) i) = 0,
+
+  rw [this, add_zero], apply finset.sum_congr rfl,
```

Comment:
If we write this in terms of `proj`, then there won't be any coercion going on which makes proof easier to write.

Reply:
You can always reintroduce `proj` inside the proof by rewriting `proj_apply` backwards.